### DISCUSSION TITLE
General Diffusers Q&A (Non-SDXL)

### DISCUSSION THREAD

**Discussion by Aptronymist** [vladmandic/sdnext#1635] [Q&A]
Starting up a new Q&A here as you can see, this is devoted to the Huggingface Diffusers backend itself, using it for general image generation. All SDXL questions should go in the SDXL Q&A.

**Comment by Threnos** [comment]
Came across this topic looking for an information about Diffusers, as I know nothing about them at all.

**Comment by setothegreat** [comment]
Interested in using Diffusers for the benefits that custom pipelines bring, but there's not much documentation when it comes to setting it up.

When I try to switch to Diffusers in the settings menu and then either reload or restart the server, I get the following error:

`Error listing diffusers: F:\Diffusion\automaticNEW\automatic\models\Diffusers Cache directory not found: F:\Diffusion\automaticNEW\automatic\models\Diffusers. Please use `cache_dir` argument or set `HUGGINGFACE_HUB_CACHE` environment variable.`

Issue is I'm not really sure how to use the "cache_dir" argument, nor how to set the "HUGGINGFACE_HUB_CACHE" environment variable. I'm also not sure what exactly it is I'm supposed to be putting in the Diffusers directory if I am specifically looking to use standard Stable Diffusion 1.5 models with it.

Any advice?

**Reply by vladmandic** [reply]
you need to start with `--backend diffusers` once so settings get intitialized. after that you can switch using ui. i'll address that tomorrow.

**Reply by setothegreat** [reply]
> you need to start with `--backend diffusers` once so settings get intitialized. after that you can switch using ui. i'll address that tomorrow.

Thank you for the quick response. I believe I am doing this, though I may have setup the webui.bat file incorrectly. This is what the relevant section currently looks like:

```
@echo off

if not defined PYTHON (set PYTHON=python)
if not defined VENV_DIR (set "VENV_DIR=%~dp0%venv")
set ERROR_REPORTING=FALSE
set COMMANDLINE_ARGS=--backend diffusers
mkdir tmp 2>NUL
```

Reason I think I might have set it up incorrectly is because it seems to list the argument as two separate arguments in the console:

`13:27:42-420526 INFO     Server arguments: ['--backend', 'diffusers']`

That being said, when I entered an incorrect argument earlier it threw up a fatal exception, so idk.
Either way, I still get the error after trying to launch and relaunch the .bat file with the above argument, so clearly I'm doing something wrong.

**Reply by setothegreat** [reply]
> > you need to start with `--backend diffusers` once so settings get intitialized. after that you can switch using ui. i'll address that tomorrow.
> 
> Thank you for the quick response. I believe I am doing this, though I may have setup the webui.bat file incorrectly. This is what the relevant section currently looks like:
> 
> ```
> @echo off
> 
> if not defined PYTHON (set PYTHON=python)
> if not defined VENV_DIR (set "VENV_DIR=%~dp0%venv")
> set ERROR_REPORTING=FALSE
> set COMMANDLINE_ARGS=--backend diffusers
> mkdir tmp 2>NUL
> ```
> 
> Reason I think I might have set it up incorrectly is because it seems to list the argument as two separate arguments in the console:
> 
> `13:27:42-420526 INFO Server arguments: ['--backend', 'diffusers']`
> 
> That being said, when I entered an incorrect argument earlier it threw up a fatal exception, so idk. Either way, I still get the error after trying to launch and relaunch the .bat file with the above argument, so clearly I'm doing something wrong.

Seems to be generating images just fine despite the error still being listed in the console, so I guess the fix worked lol. Thank you.

**Reply by vladmandic** [reply]
Put the arguments in quotes.
And I **strongly** advisevnot to modify webui.bat.
Its not related to this, but there are tons of reasons for it plus any number of documented alternatives.

**Reply by setothegreat** [reply]
> Put the arguments in quotes. And I **strongly** advisevnot to modify webui.bat. Its not related to this, but there are tons of reasons for it plus any number of documented alternatives.

Will do, and which alternative would you specifically suggest for setting up these arguments? I had previously always been under the impression that these arguments were to go in the launch .bat file since that's how it was done in other repos, so I wasn't aware that there was anywhere else you could specify them.

**Reply by vladmandic** [reply]
check faq in #1011

**Comment by setothegreat** [comment]
Pretty impressed with Diffusers so far; not only does the performance seem to be drastically higher (more than 2x what I get with the standard backend) but it also seems to produce higher quality results based on my testing so far.

Now it is obviously still in an early state as can be seen by the fact that a lot of the functionality available to the standard backend is as of yet not available, but I am curious if a way to implement custom community pipelines is possible in this current implementation. [The official documentation regarding them](https://github.com/huggingface/diffusers/tree/main/examples/community) does not seem particularly helpful as the implementation through the UI is drastically different, and looking through the scripts throughout the Diffusers folder in the venv directory doesn't seem to turn up anything helpful from what I could find.

Is there currently a way to do this, or should I just wait for further developments?

**Reply by vladmandic** [reply]
just to clarify, what would you consider a custom community pipeline in your case?

**Reply by setothegreat** [reply]
> just to clarify, what would you consider a custom community pipeline in your case?

Not entirely sure what you mean, but my primary interest in Diffusers was making use of the "CLIP Guided Stable Diffusion" and "Long Prompt Weighting Stable Diffusion" community pipelines, since CLIP Guided Stable Diffusion can supposedly allow for far higher quality generations at the cost of higher VRAM consumption, and Long Prompt Weighting Stable Diffusion can supposedly remove token limits from prompts without the drawbacks that going over the 75 token limit in A1111 repos tend to have.

This is how they outline using them in the Github documentation I provided earlier:

> To load a custom pipeline you just need to pass the custom_pipeline argument to DiffusionPipeline, as one of the files in diffusers/examples/community. Feel free to send a PR with your own pipelines, we will merge them quickly.
> 
> pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", custom_pipeline="filename_in_the_community_folder")

**Comment by TeutonJon78** [comment]
Is the diffuser backend going to be preferred backend in general going forward? It seems like it's the "new method" for all diffusions on Huggingface, especially since it supports loading 1.5 models (even if not the all the extensions yet).

**Reply by vladmandic** [reply]
not sure about preferred as there are a lot of standard workflows that are important and not supported (yet) with diffusers. so i'm trying to have both side-by-side as much as possible for now.
